### ISSUE TITLE
Option clash for package hyperref.

### ISSUE THREAD

**Issue by homocomputeris** [mlhetland/skeldoc.sty#3]
Hei!

I'd like to use skeldoc and highlight links in the PDF.
However, this code does not compile because of `Option clash for package hyperref.`:
```
\documentclass[]{article}

\author{Name}
\title{Title}
\usepackage{skeldoc}
\usepackage[colorlinks]{hyperref}
\begin{document}
Text.
\end{document}
```

While this one does:
```
\documentclass[]{article}

\author{Name}
\title{Title}
\usepackage{skeldoc}
\usepackage{hyperref}
\begin{document}
Text.
\end{document}
```

**Comment by mlhetland** [comment]
Hi!

Thanks for the [MWE](https://en.wikipedia.org/wiki/Minimal_working_example)s!-)

Right, this is because `skeldoc` already uses `hyperref`, and supplies its own options to it. When you then use `hyperref` afterward, using other options (in this case `colorlinks`), you get the conflict error.

Instead, you could use `\hypersetup` to provide `hyperref` with other options, possibly overriding those used by `skeldoc`. That is, you could do something like:

```latex
\documentclass[]{article}

\author{Name}
\title{Title}
\usepackage{skeldoc}
\usepackage{hyperref} % Not really needed
\hypersetup{colorlinks}
\begin{document}
Text.
\end{document}
```

Note that while you don't need `\usepackage{hyperref}` (because it's already used by `skeldoc`), it might be useful to keep it, as you'll probably be removing hte `\usepackage{skeldoc}` command once the document is finished :-)

Just let me know if this ends up not doing what you want. I'll leave the issue open for now – I might want to address this either in the code some way, or perhaps in the docs.

**Comment by mlhetland** [comment]
I was a bit quick, there xD I remember now that there actually *is* a mechanism for overriding the options used by `skeldoc`; it's the `hypersetup` key. The default is `hidelinks`, i.e.,

```latex
\usepackage{skeldoc}
```

is equivalent to:

```latex
\usepackage[hypersetup=hidelinks]{skeldoc}
```

So you can just use this instead:

```latex
\usepackage[hypersetup=colorlinks]{skeldoc}
```

I guess it might be useful to have an option to turn off the importing of `hyperref` in `skeldoc` altogether, so you could use `\usepackage[colorlinks]{hyperref}` without getting any trouble (so that your code, apart from the `skeldoc` import, can remain as-is once `skeldoc` is gone – which it also can with the `\hypersetup` command, of course).

So, yeah, there is already support for configuring the use of `hyperref`. I guess that fixes the problem, but I'll still leave the issue open – because I'd really like to make it possible to just use options directly with `\usepackage[…]{hyperref}`, like you did. Either by switching off the import in `skeldoc` (which should be possible for other packages imported by `skeldoc`, too, for similar reasons) or perhaps in some other way. I'll look into it. But at least your particular problem should be solved, I hope?

**Comment by homocomputeris** [comment]
Thanks!
Is it possible to pass several options to skeldoc? `pdfusetitle` cannot be passed to `\hypersetup`. Can I use \usepackage[hypersetup=hidelinks,pdfusetitle]{skeldoc}?

By the way, I believe that there otheer packages that load hyperref, for example, glossaries(-extra). But they won't cause clash in this MWE. How do they avoid it?

**Comment by mlhetland** [comment]
OK, I was clearly not entirely awake when I answered you yesterday xD The `skeldoc` package doesn't take any options to `\usepackage` (partly because of some expansion issues that occur there that I'd rather not deal with). Instead, use `\skelset` to supply these options. When it comes to the `hypersetup`, that is stored and then executed automatically in a `\hypersetup` command at `\begin{document}`.

You can indeed pass multiple arguments to an option – just enclose them in braces, like `hypersetup={foo,bar}`. Whatever is found in the braces will simply replace the default used by `\skeldoc`. This will just be passed to `\hypersetup`, however. I guess (maybe) it *ought* to be used when importing `hyperref`, since there are options like `pdfusetitle` that only work there.

As for how `glossaries` and `glossaries-extra` do it: Without having studied the code in detail, it *seems* they avoid the issue by simply not importing `hyperref`. They *use* it (if it's available), but you have to do the importing. Or so it seems. That might be a sensible solution for `skeldoc` too, I guess; i.e., if you import `hyperref`, the numbers be come links (for example).

To solve your problem, specifically, I guess you *could* use [`pdftitle` and `pdfauthor`](https://tex.stackexchange.com/a/17254) instead (i.e., with `\@title` and `\@author` as arguments).

A better solution is probably to use something like the following:
```latex
\PassOptionsToPackage{colorlinks,pdfusetitle}{hyperref}
\usepackage{skeldoc}
\usepackage{hyperref} % optional
```
The last `\usepackage` is just so that if you comment out/delete the `skeldoc` import (once you're done filling out your document), you'll still get `hyperref`, with the options you passed to it.

This isn't entirely functional, however. Ideally, you should also be able to add `\skelset{hypersetup=}` to empty out the setup there, but currently `skeldoc` will happily call `\hypersetup` with an empty token list, and it seems `\hypersetup` doesn't like that (even though `\hypersetup{}` works just fine; I think the empty token list is interpreted as a key name – I guess that's essentially a bug in how I call `\hypersetup`, then).

You can always just add `\skelset{hypersetup=colorlinks}` or something, just to put something in there, and it should work. If you have that in `\PassOptionsToPackage` as well, you'll get a warning. You could always delete that option from there, but then of course you no longer get the right setup if you just delete the `skeldoc` stuff (which I'm just assuming is a natural thing to do once it's no longer needed).

So … I think this is a functioning workaround, though a better solution would be, well … better :-)

```latex
\documentclass{article}

\author{Name}
\title{Title}
\PassOptionsToPackage{pdfusetitle}{hyperref}
\usepackage{skeldoc}
\skelset{hypersetup=colorlinks}
\begin{document}
Text.
\end{document}
```

**Comment by wenlzhang** [comment]
Thanks for the clarification.

> Does this work for you? Could you perhaps. see if you could update the example to more closely mirror your document, so that you get the error you describe?

Yes, the provided example document works for me as well.

If `\skelpar{Refer to \href{Link}{Text} for more details.}` is replaced by using an Obsidian URI, e.g. the following, then it gives the error as described above.
```
\skelpar{Refer to \href{obsidian://open?vault=my%20vault&file=my%20note.md}{Text} for more details.}
```

**Comment by mlhetland** [comment]
The same issue occurs with `enumitem` (and possibly other packages; I haven't checked). I guess a more general solution is needed, here.
